File: csharp/dnd-character/DndCharacter.cs
public class DndCharacter
{
    public int Strength { get; }
    public int Dexterity { get; }
    public int Constitution { get; }
    public int Intelligence { get; }
    public int Wisdom { get; }
    public int Charisma { get; }
    public int Hitpoints { get; }
    private static readonly Random _random = new();
    private DndCharacter()
    {
        Strength = Ability();
        Dexterity = Ability();
        Constitution = Ability();
        Intelligence = Ability();
        Wisdom = Ability();
        Charisma = Ability();
        Hitpoints = 10 + Modifier(Constitution);
    }
    public static int Modifier(int score)
    => (int)Math.Floor((score - 10) / 2.0);

    public static int Ability() 
    => Enumerable.Range(0, 4)
                         .Select(_ => _random.Next(1, 7))
                         .OrderBy(score => score)
                         .Skip(1)
                         .Sum();
    
    public static DndCharacter Generate()
    => new();
}

File: javascript/dnd-character/dnd-character.js
//
// This is only a SKELETON file for the 'D&D Character' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const abilityModifier = (score) => {
  if (score < 3) throw new Error(`Ability scores must be at least 3`);
  if (score > 18) throw new Error(`Ability scores can be at most 18`);
  return Math.floor((score - 10 ) / 2);
};

export class Character {
  constructor(){
    this._strength = Character.rollAbility();
    this._dexterity = Character.rollAbility();
    this._constitution = Character.rollAbility();
    this._intelligence = Character.rollAbility();
    this._wisdom = Character.rollAbility();
    this._charisma = Character.rollAbility();
    this._hitpoints = 10 + abilityModifier(this._constitution);
  }

  static rollAbility() {
    const rolls = Array.from({ length: 4}, () => Math.floor(Math.random() * 6) + 1);
    return rolls.sort((a, b) => b - a).slice(0, 3).reduce((acc, cur) => acc + cur, 0);
  }

  get strength() { return this._strength }
  get dexterity() { return this._dexterity }
  get constitution() { return this._constitution }
  get intelligence() { return this._intelligence }
  get wisdom() { return this._wisdom }
  get charisma() { return this._charisma }
  get hitpoints() { return this._hitpoints }
}
